File: 712-MinimumAsciiDeleteSumForTwoStrings/712-MinimumAsciiDeleteSumForTwoStrings.py
# Last updated: 12/6/2025, 5:47:31 am
alp = {'a': 97, 'b': 98, 'c': 99, 'd': 100, 'e': 101, 'f': 102, 'g': 103, 'h': 104, 'i': 105, 'j': 106, 'k': 107, 'l': 108, 'm': 109, 'n': 110, 'o': 111, 'p': 112, 'q': 113, 'r': 114, 's': 115, 't': 116, 'u': 117, 'v': 118, 'w': 119, 'x': 120, 'y': 121, 'z': 122}

class Solution:
    def minimumDeleteSum(self, s1: str, s2: str) -> int:
        m = len(s1)
        n = len(s2)
        tot = 0
        for s in s1:
            tot += alp[s]
        for s in s2:
            tot += alp[s]
        dp = [[tot for _ in range(n+1)] for _ in range(m+1)]

        for i in range(1, m+1):
            for j in range(1, n+1):
                if s1[i-1] == s2[j-1]:
                    dp[i][j] = dp[i-1][j-1] - (alp[s1[i-1]] * 2)
                else:
                    dp[i][j] = min(dp[i-1][j], dp[i][j-1])
        
        return dp[-1][-1]
        

File: 712-MinimumASCIIDeleteSumforTwoStrings/712-MinimumASCIIDeleteSumforTwoStrings.cpp
// Last updated: 1/10/2026, 3:55:57 PM
1
2int dp[1001][1001];
3int inf = 1e9 * 2;
4
5int rec(int i, int j, int &m, int &n, string &s1, string &s2,
6        vector<int> &prefix_1, vector<int> &prefix_2) {
7  if (dp[i][j] != -1)
8    return dp[i][j];
9
10  if (i == m)
11    return prefix_2[n] - prefix_2[j];
12  if (j == n)
13    return prefix_1[m] - prefix_1[i];
14
15  int ans = min(rec(i + 1, j, m, n, s1, s2, prefix_1, prefix_2) + s1[i],
16                rec(i, j + 1, m, n, s1, s2, prefix_1, prefix_2) + s2[j]);
17  if (s1[i] == s2[j])
18    ans = min(ans, rec(i + 1, j + 1, m, n, s1, s2, prefix_1, prefix_2));
19
20  dp[i][j] = ans;
21  return ans;
22}
23
24class Solution {
25public:
26  int minimumDeleteSum(string s1, string s2) {
27    memset(dp, -1, sizeof(dp));
28    int m = s1.length();
29    int n = s2.length();
30
31    vector<int> prefix_1(m + 1, 0), prefix_2(n + 1, 0);
32    int val = 0;
33    for (int i = 0; i < m; i++) {
34      prefix_1[i] = val;
35      val += s1[i];
36    }
37    prefix_1[m] = val;
38
39    val = 0;
40    for (int i = 0; i < n; i++) {
41      prefix_2[i] = val;
42      val += s2[i];
43    }
44    prefix_2[n] = val;
45
46    return rec(0, 0, m, n, s1, s2, prefix_1, prefix_2);
47  }
48};
49